### PR TITLE
api: add user.get, register user service

### PR DESCRIPTION
**Pull request by sorja**
## Description
- [ ] Implement AuthMiddleware in ArenaServer.init
- [x] Implement TableUser + UserService.get and register UserService in ArenaServer.init (registerServices)
- [ ] Merge authMiddleware with accessControlMiddleware
- [ ] Implement Google login + local login strategies

## Types of changes

- [ ] Docs change
- [ ] Refactoring
- [ ] Dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How has this been tested

No testing

#### Do you consider this PR needs further testing?

- [ x] No
- [ ] Yes

### CODE REVIEW

**Comment by minotogna** [comment]
Task linked: [CU-h10vj4 AuthMiddleware](https://app.clickup.com/t/h10vj4)

File: src/db/table/table.ts [hunk 1]
```diff
@@ -2,17 +2,17 @@ import { SQLs } from '../sql'
 
 export abstract class Table {
   readonly schema: string
-  readonly name: string
+  readonly tableName: string
```
Comment (sorja) [inline src/db/table/table.ts:5]: Changed to avoid conflict if `name` is used

**Comment by SteRiccio** [comment]
:+1:

File: src/repository/user/get.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  const table = new TableUser()
+  const selectFields = [table.uuid, table.name, table.email, table.prefs, table.status, table.props]
+
+  // TODO: Support fetching by email [and password]
+  // // Using email and password to fetch user
+  // // Do not include password by default
+  // if (!userUuid) selectFields.push(password)
+
+  const where = `user.uuid = '${userUuid}'`
```
Comment (minotogna) [inline src/repository/user/get.ts:21]: this should be table.uuid =

File: src/db/table/schemaPublic/user.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import { TableSchemaPublic } from './tableSchemaPublic'
+
+export class TableUser extends TableSchemaPublic {
+  readonly uuid: Column = new Column(this, 'uuid', ColumnType.uuid)
+  readonly name: Column = new Column(this, 'name', ColumnType.varchar)
+  readonly email: Column = new Column(this, 'email', ColumnType.varchar)
+  readonly password: Column = new Column(this, 'password', ColumnType.varchar)
+  readonly prefs: Column = new Column(this, 'prefs', ColumnType.jsonb)
+  readonly profile_picture: Column = new Column(this, 'profile_picture', ColumnType.bytea)
+  // TODO: This should be enum: user_status
```
Comment (minotogna) [inline src/db/table/schemaPublic/user.ts:11]: not now... leave it as ColumnType.varchar

File: src/db/table/schemaPublic/user.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  readonly name: Column = new Column(this, 'name', ColumnType.varchar)
+  readonly email: Column = new Column(this, 'email', ColumnType.varchar)
+  readonly password: Column = new Column(this, 'password', ColumnType.varchar)
+  readonly prefs: Column = new Column(this, 'prefs', ColumnType.jsonb)
+  readonly profile_picture: Column = new Column(this, 'profile_picture', ColumnType.bytea)
+  readonly status: Column = new Column(this, 'status', ColumnType.varchar)
+  readonly props: Column = new Column(this, 'props', ColumnType.jsonb)
+
+  constructor() {
+    super('user')
```
Comment (minotogna) [inline src/db/table/schemaPublic/user.ts:15]: user might give problems in queries.
it must be escaped with ""
e.g.
```
super('"user"')
```

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# top-most EditorConfig file
```
Comment (minotogna) [inline .editorconfig:1]: we don't need this

File: package.json [hunk 6]
```diff
@@ -36,7 +36,11 @@
   },
   "dependencies": {
     "@godaddy/terminus": "^4.7.1",
-    "@openforis/arena-core": "^0.0.55",
+    "@openforis/arena-core": "^0.0.59",
+    "@types/bcryptjs": "^2.4.2",
+    "@types/passport": "^1.0.6",
+    "@types/passport-local": "^1.0.33",
```
Comment (minotogna) [inline package.json:42]: @types go to dev dependencies

File: src/db/table/schemaPublic/user.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { Column, ColumnType } from '../../column'
+import { TableSchemaPublic } from './tableSchemaPublic'
+
+export class TableUser extends TableSchemaPublic {
+  readonly uuid: Column = new Column(this, 'uuid', ColumnType.uuid)
+  readonly name: Column = new Column(this, 'name', ColumnType.varchar)
+  readonly email: Column = new Column(this, 'email', ColumnType.varchar)
+  readonly password: Column = new Column(this, 'password', ColumnType.varchar)
+  readonly prefs: Column = new Column(this, 'prefs', ColumnType.jsonb)
+  readonly profile_picture: Column = new Column(this, 'profile_picture', ColumnType.bytea)
```
Comment (minotogna) [inline src/db/table/schemaPublic/user.ts:10]: property name profile_picture should be profilePicture

File: src/repository/user/get.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { BaseProtocol, DB, SqlSelectBuilder, TableUser } from '../../db'
+import { User } from '@openforis/arena-core'
+
+/**
+ * Returns a user by id.
+ *
+ * @param client - Database client.
+ */
+type getOptionsType = { userUuid: string } | { email: string }
```
Comment (minotogna) [inline src/repository/user/get.ts:9]: password is missing

File: src/repository/user/get.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  const table = new TableUser()
+  const selectFields = [table.uuid, table.name, table.email, table.prefs, table.status, table.props]
+  let where = ''
+
+  if ('email' in options) {
+    where = `table.email = '${options.email}'`
+    // Using email and password to fetch user
+    // Do not include password by default
+    selectFields.push(table.password)
```
Comment (minotogna) [inline src/repository/user/get.ts:21]: why selecting password ?

File: src/repository/user/get.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  } else {
+    where = `table.uuid = '${options.userUuid}'`
```
Comment (minotogna) [inline src/repository/user/get.ts:23]: do not hardcode field names. use always table columns. this query won't work this way

File: src/repository/userResetPassword/index.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export const UserResetPasswordRepository = {
+  // @ts-ignore TODO
+  // eslint-disable-next-line
+  existsResetPasswordValidByUserUuid: (options: { userUuid: string }): boolean => {},
```
Comment (minotogna) [inline src/repository/userResetPassword/index.ts:4]: call this method simply hasValidResetPassword

File: src/server/arenaApp.ts [hunk 12]
```diff
@@ -1,9 +1,19 @@
...
   express: Express
   session: RequestHandler
 }
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace Express {
+    // Express.User is defined as {}
+    // We extend it globally to be arena-core/User
+    // @ts-ignore
```
Comment (minotogna) [inline src/server/arenaApp.ts:16]: why ts-ignore ?

File: src/server/arenaApp.ts [hunk 13]
```diff
@@ -1,9 +1,19 @@
...
 /**
  * Express app wrapper.
  */
 export interface ArenaApp {
   express: Express
   session: RequestHandler
 }
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-namespace
```
Comment (minotogna) [inline src/server/arenaApp.ts:12]: this is not good. better to add 
```
rules: {
  "@typescript-eslint/no-namespace": "off"
}
```
?

File: src/server/middleware/authentication.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    const expired = !(await UserResetPasswordRepository.existsResetPasswordValidByUserUuid({ userUuid: user.uuid }))
+    userUpdated = {
+      ...userUpdated,
+      invitation: {
+        expired,
+      },
+    }
+  }
+
+  // @ts-ignore fixed by implementing repository method
```
Comment (minotogna) [inline src/server/middleware/authentication.ts:39]: why ts-ignore ?

File: src/server/middleware/authentication.ts [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+import { Express } from 'express'
+import { FieldValidators, User, UserStatus, Validator } from '@openforis/arena-core'
+
+import { ExpressInitializer } from '../expressInitializer'
+import { UserServiceServer } from '../../service'
+import { UserResetPasswordRepository, AuthGroupRepository } from '../../repository'
+
+const comparePassword = bcrypt.compareSync
+
+// TODO: Handle these separately
```
Comment (minotogna) [inline src/server/middleware/authentication.ts:13]: what is this TODO?

File: src/server/middleware/authentication.ts [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  if (user.status === UserStatus.ACCEPTED) {
+    sendUser(user)
+  }
+
+  sendError(ValidationMessages.passwordChangeRequired)
+}
+
+const LocalStrategy = passportLocal.Strategy
```
Comment (minotogna) [inline src/server/middleware/authentication.ts:83]: why this? cant you just use passportLocal.Strategy? or import Strategy as LocalStrategy ?

File: src/utils/requests.ts [hunk 17]
```diff
@@ -1,7 +1,7 @@
 import { Request } from 'express'
 import { User } from '@openforis/arena-core'
 
-type ArenaRequest = { user: User } & Request
+export type ArenaRequest = { user: User } & Request
```
Comment (minotogna) [inline src/utils/requests.ts:4]: do we still need this??

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/server/middleware/authentication.ts [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        expired,
+      },
+    }
+  }
+
+  // @ts-ignore fixed by implementing repository method
+  return userUpdated
+}
+
+const findUserByEmailAndPassword = async (email: string, password: string) => {
```
Comment (SteRiccio) [inline src/server/middleware/authentication.ts:43]: Don't you need to specify the return type?

Comment (sorja) [reply src/server/middleware/authentication.ts:43]: The return type is inherited from `_initializeUser`

File: src/server/middleware/authentication.ts (hunk 14 above)
Comment (sorja) [reply src/server/middleware/authentication.ts:39]:
```
Type '{ authGroups: void; email: string; hasProfilePicture: boolean; name: string; prefs?: UserPrefs | undefined; status: UserStatus; password?: string | undefined; invitation?: UserInvitation | undefined; uuid: string; props: UserProps; validation?: Validation | undefined; }' is not assignable to type 'User'.
  Types of property 'authGroups' are incompatible.
    Type 'void' is not assignable to type 'AuthGroup[] | undefined'.ts(2322)
    ```

File: src/server/middleware/authentication.ts (hunk 15 above)
Comment (sorja) [reply src/server/middleware/authentication.ts:13]: Should we keep those messages here or move them to core?

File: src/server/arenaApp.ts (hunk 13 above)
Comment (sorja) [reply src/server/arenaApp.ts:12]: Are we allowed to use it elsewhere also in this case?

File: src/repository/user/get.ts (hunk 9 above)
Comment (sorja) [reply src/repository/user/get.ts:21]: To compare with bcrypt, ilmentation equals Arena's implementation
`comparePassword(password, user.password)`

File: src/repository/user/get.ts (hunk 8 above)
Comment (sorja) [reply src/repository/user/get.ts:9]: password is not used as param

File: src/repository/user/get.ts (hunk 9 above)
Comment (minotogna) [reply src/repository/user/get.ts:21]: it must be selected when password is passed as parameter.
if password is passed as parameter, the check should be done here, not elsewhere

File: src/server/arenaApp.ts (hunk 13 above)
Comment (minotogna) [reply src/server/arenaApp.ts:12]: you can use export instead of namespace.

File: src/server/middleware/authentication.ts (hunk 15 above)
Comment (minotogna) [reply src/server/middleware/authentication.ts:13]: I didn't understand what is the TODO

File: src/server/middleware/authentication.ts (hunk 14 above)
Comment (minotogna) [reply src/server/middleware/authentication.ts:39]: and there's no way to fix this?

File: src/server/middleware/authentication.ts (hunk 18 above)
Comment (minotogna) [reply src/server/middleware/authentication.ts:43]: better to be explicit

File: src/server/middleware/authentication.ts (hunk 15 above)
Comment (sorja) [reply src/server/middleware/authentication.ts:13]:
```
// TODO: Handle these separately
const ValidationMessages = {
  emailInvalid: 'validationErrors.user.emailInvalid',
  userNotFound: 'validationErrors.user.userNotFound',
  passwordChangeRequired: 'validationErrors.user.passwordChangeRequired',
}
```

Where should these be?

File: src/server/middleware/authentication.ts (hunk 14 above)
Comment (sorja) [reply src/server/middleware/authentication.ts:39]: The ts-ignore comment has `fixed by implementing repository method`

File: src/repository/user/get.ts (hunk 8 above)
Comment (minotogna) [reply src/repository/user/get.ts:9]: pass it here

File: src/server/arenaApp.ts (hunk 12 above)
Comment (sorja) [reply src/server/arenaApp.ts:16]: As in comment:
`     // Express.User is defined as {} `
```
Duplicate identifier 'User'.ts(2300)
index.d.ts(20, 19): 'User' was also declared here
```

File: src/server/middleware/authentication.ts [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    express.use(passport.initialize())
+
+    express.use(passport.session())
+
+    passport.use(localStrategy)
+
+    passport.serializeUser((user, done) => done(null, user?.uuid))
+
+    passport.deserializeUser(async (userUuid: string, done) => {
+      const user = await UserServiceServer.get({ userUuid })
```
Comment (minotogna) [inline src/server/middleware/authentication.ts:59]: use ServiceRegistry to get UserService

**Review by minotogna — CHANGES_REQUESTED** [review]

**Review by SteRiccio — APPROVED** [review]

**Review by minotogna — APPROVED** [review]

**Pull request merged**